### PR TITLE
Add reference numbers to email

### PR DESCRIPTION
**Pull request by DavidBiddle**
### What problem does this pull request solve?

Trello card: https://trello.com/c/dm8jKunt/1361-generate-a-reference-for-form-submissions-email-work

Adds two new variables to the Notify personalisation for the submission email (sent to processors) and confirmation email (sent to form fillers, if requested):

- include_submission_reference: this is a boolean controlled by the `reference_numbers_enabled` feature flag.
- submission_reference: the reference number itself.

This means that once this is merged and deployed, we can add the variable to the Notify templates in the format `((INCLUDE_SUBMISSION_REFERENCE??Your reference number is ))((SUBMISSION_REFERENCE))`. When the feature flag is off, this will return nothing, so the email will be the same as it was previously. But when the flag is on, it will return `Your reference is R34R3NC3`.

### Things to consider when reviewing

- Ensure that you consider the wider context.
- Does it work when run on your machine?
- Is it clear what the code is doing?
- Do the commit messages explain why the changes were made?
- Are there all the unit tests needed?
- Has all relevant documentation been updated?

### CODE REVIEW

File: app/mailers/form_submission_mailer.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 class FormSubmissionMailer < GovukNotifyRails::Mailer
-  def email_completed_form(title:, text_input:, reference:, preview_mode:, timestamp:, submission_email:)
+  def email_completed_form(title:, text_input:, reference:, preview_mode:, timestamp:, submission_email:, submission_reference:)
```
Comment (chao-xian) [inline app/mailers/form_submission_mailer.rb:2]: There's an opportunity here to refactor https://refactoring.guru/introduce-parameter-object

**Review by chao-xian — COMMENTED** [review]
Just a quick look but not done a proper review yet

File: spec/requests/forms/check_your_answers_controller_spec.rb [hunk 2]
```diff
@@ -85,6 +87,10 @@
       allow(context_spy).to receive(:form_submitted?).and_return(repeat_form_submission)
       context_spy
     end
+
+    # rubocop:disable RSpec/AnyInstance
```
Comment (chao-xian) [inline spec/requests/forms/check_your_answers_controller_spec.rb:91]: I'm a bit concerned about using this. There are pros and cons (see https://medium.com/beambenefits/rspec-tips-allow-any-instance-of-c11e38b098be) but I'm wondering if we need to test it in the controller. What are your thoughts?

**Review by chao-xian — COMMENTED** [review]
Just a couple thoughts. Looking good! Thanks David

File: spec/requests/forms/check_your_answers_controller_spec.rb (hunk 2 above)
Comment (chao-xian) [reply spec/requests/forms/check_your_answers_controller_spec.rb:91]: OK cool, give it a go

File: app/mailers/form_submission_mailer.rb (hunk 1 above)
Comment (DavidBiddle) [reply app/mailers/form_submission_mailer.rb]: I've done something along these lines - created a `FormSubmissionService::FormSubmissionMailerOptions` class which covers the parameters shared by this function and the one in `FormSubmissionConfirmationMailer`

File: spec/requests/forms/check_your_answers_controller_spec.rb (hunk 2 above)
Comment (DavidBiddle) [reply spec/requests/forms/check_your_answers_controller_spec.rb]: @chao-xian Done, thanks for the suggestion!

File: app/services/form_submission_service.rb [hunk 3]
```diff
@@ -5,6 +5,8 @@ def call(**args)
     end
   end
 
+  FormSubmissionMailerOptions = Data.define(:title, :preview_mode, :timestamp, :submission_reference)
```
Comment (chao-xian) [inline app/services/form_submission_service.rb:8]: I know it's a nitpick but: Can we shorten this to `MailerOptions` or just `Options` please? Given it's already namespaced in the module, it's otherwise quite long 🙏

File: app/services/form_submission_service.rb [hunk 4]
```diff
@@ -42,14 +42,12 @@ def submit_form_to_processing_team
     end
 
     unless @form.submission_email.blank? && @preview_mode
+
       mail = FormSubmissionMailer
-      .email_completed_form(title: form_title,
-                            text_input: email_body,
-                            preview_mode: @preview_mode,
+      .email_completed_form(text_input: email_body,
```
Comment (chao-xian) [inline app/services/form_submission_service.rb:47]: Not 100% sure and will need to look at it again tomorrow but feels like this could be something like:

```ruby
.email_completed_form(email, options)
```

File: app/services/form_submission_service.rb (hunk 3 above)
Comment (DavidBiddle) [reply app/services/form_submission_service.rb]: Good shout, have gone with `MailerOptions`

**Comment by DavidBiddle** [comment]
Have pushed an extra commit to resolve an ambiguity that was bugging me.

File: app/services/form_submission_service.rb (hunk 4 above)
Comment (DavidBiddle) [reply app/services/form_submission_service.rb:47]: @chao-xian what would `email` contain here?

Comment (DavidBiddle) [reply app/services/form_submission_service.rb:47]: I'm inclined to do this in a subsequent PR but yes, something like that would make sense

**Review by chao-xian — APPROVED** [review]
Thanks David!
